### PR TITLE
Fix ``Ledger`` test so they pass regardless of system locale

### PR DESCRIPTION
**Pull request by manumafe98**
# pull request

Related issue: [#2635](https://github.com/exercism/java/issues/2635)

Reviewer Resources:

[Track Policies](https://github.com/exercism/java/blob/main/POLICIES.md#event-checklist)

### CODE REVIEW

**Comment by manumafe98** [comment]
The only issue I could see with this fix is that the students think that part of the code needs to be reformatted, and then do not know how to go back to the previous working state. 

Also maybe this need to be another discussion but a lot of students (checking on community solutions) send the exercise raw without changes, maybe this reformat style are not that exciting? or maybe we could make a check so at least they need to make a change to the original code? 

Apart from that ledger has the same implementation in reference as well in the code given to the students, which I guess is not the idea, comparing to the ``Markdown`` reformat exercise.

**Comment by sanderploegsma** [comment]
> The only issue I could see with this fix is that the students think that part of the code needs to be reformatted, and then do not know how to go back to the previous working state.

Yes, ideally you should come up with a solution that works regardless of how students change the implementation. One way I could think of is to temporarily change the JVM locale to `en-US` from the test suite, perhaps using an `@Rule` like the one mentioned here: https://stackoverflow.com/a/76430798

> Also maybe this need to be another discussion but a lot of students (checking on community solutions) send the exercise raw without changes, maybe this reformat style are not that exciting? or maybe we could make a check so at least they need to make a change to the original code?
> 
> Apart from that ledger has the same implementation in reference as well in the code given to the students, which I guess is not the idea, comparing to the `Markdown` reformat exercise.

Yes I think this exercise could be improved a bit. If you have any ideas feel free to drop them in an issue or on the forum, perhaps there are others who want to help out 👍

**Comment by manumafe98** [comment]
> > The only issue I could see with this fix is that the students think that part of the code needs to be reformatted, and then do not know how to go back to the previous working state.
> 
> Yes, ideally you should come up with a solution that works regardless of how students change the implementation. One way I could think of is to temporarily change the JVM locale to `en-US` from the test suite, perhaps using an `@Rule` like the one mentioned here: https://stackoverflow.com/a/76430798
> 
> > Also maybe this need to be another discussion but a lot of students (checking on community solutions) send the exercise raw without changes, maybe this reformat style are not that exciting? or maybe we could make a check so at least they need to make a change to the original code?
> > Apart from that ledger has the same implementation in reference as well in the code given to the students, which I guess is not the idea, comparing to the `Markdown` reformat exercise.
> 
> Yes I think this exercise could be improved a bit. If you have any ideas feel free to drop them in an issue or on the forum, perhaps there are others who want to help out 👍

I will try to implement that Rule, so in the meantime works as expected until we found how we can improve this kind of exercises!

**Comment by manumafe98** [comment]
@sanderploegsma I have a couple o questions, I rebased main into this branch and now seems to have a ton of changes, i should open a new one? Also how do you download the logs of the pipeline?

image

Cause i've done that from the image above, but I do not find them helpful.

Also the annotation locally seems to work ok, the tests now work correctly.

**Comment by sanderploegsma** [comment]
> @sanderploegsma I have a couple o questions, I rebased main into this branch and now seems to have a ton of changes, i should open a new one? 

Should be fixed now, I clicked the Update Branch button on GitHub. 

> Also how do you download the logs of the pipeline?

The logging output from the test runner isn't helpful indeed. However, when any tests fail it should also store an artifact that you can download. It contains the JSON output of the test runner for each exercise. That should contain information about what is going on.

**Comment by sanderploegsma** [comment]
The test runner is a Docker container that runs without access to the internet for security purposes. This means that you can't add dependencies to an exercise without bundling them in the test runner.

In practice we only do this for dependencies that are really required for the exercise to work (like a JSON library in the rest-api exercise).

In this case, there is an alternative way to achieve the same without installing extra dependencies by using JUnit's extension model. For example, these extension points might work: https://junit.org/junit5/docs/current/user-guide/#extensions-lifecycle-callbacks-before-after-execution. It could work similar to the `@Rule` approach for JUnit 4.

**Comment by manumafe98** [comment]
> The test runner is a Docker container that runs without access to the internet for security purposes. This means that you can't add dependencies to an exercise without bundling them in the test runner.
> 
> In practice we only do this for dependencies that are really required for the exercise to work (like a JSON library in the rest-api exercise).

Good to know! thanks!


> In this case, there is an alternative way to achieve the same without installing extra dependencies by using JUnit's extension model. For example, these extension points might work: https://junit.org/junit5/docs/current/user-guide/#extensions-lifecycle-callbacks-before-after-execution. It could work similar to the `@Rule` approach for JUnit 4.

Sure, I will give it a try then, let me read that documentation and see if I can figure it out how to do it!

File: exercises/practice/ledger/src/test/java/LocaleExtension.java
[the remark this answers is not in the document]
Comment (manumafe98) [reply exercises/practice/ledger/src/test/java/LocaleExtension.java:1]: Done!

File: exercises/practice/ledger/build.gradle
[the remark this answers is not in the document]
Comment (manumafe98) [reply exercises/practice/ledger/build.gradle:1]: Comparing to [this one](https://github.com/exercism/java/blob/main/exercises/practice/ledger/build.gradle) I don't know what changed it's literally the same

[the remark this answers is not in the document]
Comment (sanderploegsma) [reply exercises/practice/ledger/build.gradle:1]: I think the original had a final newline at the end, and that seems to have been removed. The easiest way to revert is to run `git checkout main -- exercises/practice/ledger/build.gradle`

[the remark this answers is not in the document]
Comment (manumafe98) [reply exercises/practice/ledger/build.gradle:1]: Done! im still unsure what it was i used that command and then git diff, and i got an empty output, but well now it's like before!

**Pull request merged**